File: 74. Search a 2D Matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
             int m = matrix.size();
        int n = matrix[0].size();
        
        if(m == 0 || n == 0)
            return false;
        

        int start = 0, end = n*m - 1;
        
        while(start <= end)
        {
            int mid = start + (end - start) / 2;
	
            int index = matrix[mid/n][mid%n];
            if (target == index)
                return true;
			
            else if(target < index)
                end = mid - 1;
            else
                start = mid + 1;       
        }
        return false;

    }
};

File: 74. Search a 2D Matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        
        int totalRows = matrix.length;
        int totalCols = matrix[0].length;

        int n = totalRows * totalCols;
        int s = 0;
        int e = n-1;

        while(s <= e){
            int mid =  s + (e-s)/2;
            int rowIndex = mid/totalCols;
            int colIndex = mid % totalCols;

            if(matrix[rowIndex][colIndex] == target){
                return true;
            }
            else if(matrix[rowIndex][colIndex] > target){
                e = mid - 1;
            }
            else{
                s = mid + 1;
            }
        }
        return false;
    }
}
